File: nodejs/test/e2e/hooks.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import type { SessionHooks } from "../../src/types.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const UNSUPPORTED_SDK_HOOKS_MESSAGE = "SDK hook callbacks are no longer supported";

describe("Session hooks", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    async function expectUnsupportedHooks(hooks: SessionHooks) {
        await expect(
            client.createSession({
                onPermissionRequest: approveAll,
                hooks,
            })
        ).rejects.toThrow(UNSUPPORTED_SDK_HOOKS_MESSAGE);
    }

    const hookCases: Array<[string, SessionHooks]> = [
        ["preToolUse", { onPreToolUse: () => ({ permissionDecision: "allow" }) }],
        ["postToolUse", { onPostToolUse: () => undefined }],
        ["preToolUse denial", { onPreToolUse: () => ({ permissionDecision: "deny" }) }],
        [
            "combined preToolUse and postToolUse",
            {
                onPreToolUse: () => ({ permissionDecision: "allow" }),
                onPostToolUse: () => undefined,
            },
        ],
    ];

    it.each(hookCases)("rejects SDK callback hook %s", async (_name, hooks) => {
        await expectUnsupportedHooks(hooks);
    });
});

File: go/internal/e2e/hooks_e2e_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

const unsupportedSDKHooksMessage = "SDK hook callbacks are no longer supported"

func assertUnsupportedHooks(t *testing.T, client *copilot.Client, hooks *copilot.SessionHooks) {
	t.Helper()

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		Hooks:               hooks,
	})
	if err == nil {
		if session != nil {
			_ = session.Disconnect()
		}
		t.Fatal("expected SDK callback hooks to be rejected")
	}
	if !strings.Contains(err.Error(), unsupportedSDKHooksMessage) {
		t.Fatalf("expected unsupported hooks error, got %v", err)
	}
}

func TestHooksE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	cases := map[string]*copilot.SessionHooks{
		"preToolUse": {
			OnPreToolUse: func(copilot.PreToolUseHookInput, copilot.HookInvocation) (*copilot.PreToolUseHookOutput, error) {
				return &copilot.PreToolUseHookOutput{PermissionDecision: "allow"}, nil
			},
		},
		"postToolUse": {
			OnPostToolUse: func(copilot.PostToolUseHookInput, copilot.HookInvocation) (*copilot.PostToolUseHookOutput, error) {
				return nil, nil
			},
		},
		"preToolUse denial": {
			OnPreToolUse: func(copilot.PreToolUseHookInput, copilot.HookInvocation) (*copilot.PreToolUseHookOutput, error) {
				return &copilot.PreToolUseHookOutput{PermissionDecision: "deny"}, nil
			},
		},
		"combined preToolUse and postToolUse": {
			OnPreToolUse: func(copilot.PreToolUseHookInput, copilot.HookInvocation) (*copilot.PreToolUseHookOutput, error) {
				return &copilot.PreToolUseHookOutput{PermissionDecision: "allow"}, nil
			},
			OnPostToolUse: func(copilot.PostToolUseHookInput, copilot.HookInvocation) (*copilot.PostToolUseHookOutput, error) {
				return nil, nil
			},
		},
	}

	for name, hooks := range cases {
		t.Run("rejects SDK callback hook "+name, func(t *testing.T) {
			ctx.ConfigureForTest(t)
			assertUnsupportedHooks(t, client, hooks)
		})
	}
}
